### ISSUE TITLE
Can't set Content-Type via Response::withHeader()

### ISSUE THREAD

**Issue by burzum** [cakephp/cakephp#14211]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.8.x

### What you did

```php
	    if ($controller->getRequest()->is('json')) {
	    	return (new Response())
				->withHeader('Content-Type', 'application/json')
				->withStatus(403)
				->withStringBody(json_encode([
					'redirectUrl' => '/',
				]));
	    }
```

Doing this fixed it.

```php
	    if ($controller->getRequest()->is('json')) {
	    	return (new Response())
				->withType('application/json') // <-- See this (added)
				->withStatus(403)
				->withStringBody(json_encode([
					'redirectUrl' => '/',
				]));
	    }
```

### What happened

Got a text/html response...

### What you expected to happen

Getting an application/json type response.

If the above is intended behavior then this sucks big time. If I'm forced to use withType() for that instead of being able to use withHeader() as well, then withHeader() should throw an exception telling me that I'm not allowed to do so instead of silently overriding my set value and causing me a nice debug orgy.

**Comment by jimbo2150** [comment]
`withStatus()` calls `_setContentType()` which will overwrite the content type with the default value stored in the object's `_contentType` property along with the character set (if it exists). Using `withType()` works because it overwrites the default content type property with the new value passed in, so when `withStatus()` is called it simply overwrites the value with the same value you specified in `withType()`. Not sure it is a bug since there seems to be some white-listing (though it is hard-coded and not able to be modified). Could potentially check if there is an existing 'Content-Type' header but that would require parsing the value out (there could be a character set or potentially other value included). I think `withType()` is really the best route to isolate the type from the actual header value. There is also already a [section dealing with content types in Request/Response objects](https://book.cakephp.org/3/en/controllers/request-response.html#dealing-with-content-types) in the cookbook.

**Comment by markstory** [comment]
It seems there is some awkward temporal coupling going on as

```php
$response->withStatus(403)
  ->withHeader('Content-Type', 'application/json')
```

Wouldn't behave the same and would end up with a json content type.

> Not sure it is a bug since there seems to be some white-listing (though it is hard-coded and not able to be modified). Could potentially check if there is an existing 'Content-Type' header but that would require parsing the value out (there could be a character set or potentially other value included)

This is a potential solution. Looking back through the history of Response I found 09b3f0ff5a63a03 where the current behavior was added. The intent was to preserve clearing the content-type on 204 and 304 response codes as they have no content-type. Restoring that behavior seems possible.

**Comment by burzum** [comment]
@markstory 4.0.4? This is not going to get merged back to 3.8.x?

**Comment by markstory** [comment]
@burzum I never said that it wouldn't.  This issue is present in 4.x and 3.x. I have a pull requestnopen for 4.x that I intending backporting to 3.x as well.
